libs-scala/oracle-testing/src/main/scala/testing/oracle/OracleAround.scala

```diff
@@ -76,22 +86,76 @@ object OracleAround {
     stmt.execute(s"""GRANT EXECUTE ON SYS.DBMS_LOCK TO $name""")
     stmt.execute(s"""GRANT SELECT ON V_$$MYSTAT TO $name""")
     stmt.execute(s"""GRANT SELECT ON V_$$LOCK TO $name""")
+
+    ()
+  }
+
+  private val serverErrorsTable = "servererrors_table"
+
+  // Debugging Oracle errors is very annoying:
+  // - Oracle does not provide any details for errors (it will just print "ORA-XYZ: something went wrong")
+  // - Oracle doesn't have any built in error log
+  // - Our use of anorm and CompositeSql makes it hard to see what statements we are actually sending to the server
+  //
+  // To simplify debugging, we create a trigger that inserts all failing SQL statements into a new table
+  // Notes:
+  // - The table must not exist in the schema of the new user, as otherwise Flyway complains about a non-empty schema
+  // - In Oracle, a schema is the same as a user. We'll use the sys user to own the table and the trigger.
+  // - To debug a failing trigger, execute "select * from user_errors where type = 'TRIGGER'"
+  private def createErrorTrigger(stmt: Statement): Unit = {
+    if (!tableExists(stmt, serverErrorsTable)) {
+      logger.info(s"Creating table $serverErrorsTable and trigger")
+      stmt.execute(
+        s"create table $serverErrorsTable( usr varchar2(1000), err varchar2(1000), msg clob )"
+      )
+      stmt.execute(s"""
+        create or replace trigger servererror_trigger
+        after servererror on database
+        declare
+            l_sql_text ora_name_list_t;
+            l_n        number;
+            v_stmt     clob;
+        begin
+            l_n := ora_sql_txt( l_sql_text );
+            for i in 1 .. l_n loop
+                v_stmt := v_stmt || l_sql_text(i);
+            end loop;
+
+            insert into $serverErrorsTable values ( ora_login_user, ora_server_error(1), v_stmt );
```

Comment:
not sure what this is doing: what is the additional information what we see here besides in the exceptions?
Is it only the SQL statement itself?

Reply:
This was meant as a development time tool to simplify debugging. It only prints the SQL statement, as it was seen by the Oracle server. I was annoyed by the fact how hard it is to see the result of anorm string interpolation (for example, the `#$` operator is applied in anorm code and not our code).

I would not want to deploy this in production. Database operators have other tools to audit their Oracle database.

This is also not needed for Postgres, as Postgres includes the SQL statement in the exception message, and even points out where the error is.